core/teams/snapshot.go

```diff
@@ -0,0 +1,205 @@
+	return e.stopped
+}
+
+func (e *SnapshotEngine) StopSnapshots() {
+	e.stopped = true
+}
+
+func (e *SnapshotEngine) serialise(k string) ([]byte, error) {
+	if e.stopped {
+		return nil, nil
+	}
+
+	switch k {
+	case e.teamsKey:
+		if err := e.serialiseTeams(); err != nil {
+			return nil, err
+		}
+		return e.serialisedTeams, nil
+	case e.teamSwitchesKey:
+		if err := e.serialiseTeamSwitches(); err != nil {
+			return nil, err
+		}
+		return e.serialisedTeamSwitches, nil
+	default:
+		return nil, types.ErrSnapshotKeyDoesNotExist
+	}
+}
+
+func (e *SnapshotEngine) serialiseTeams() error {
+	teams := e.Engine.teams
+	teamsSnapshot := make([]*snapshotpb.Team, 0, len(teams))
+	for _, team := range teams {
+		refereesSnapshot := make([]string, 0, len(team.Referees))
+		for _, referee := range team.Referees {
+			refereesSnapshot = append(refereesSnapshot, string(referee))
+		}
+
+		teamSnapshot := &snapshotpb.Team{
+			Id:        string(team.ID),
+			Referrer:  string(team.Referrer),
+			Referees:  refereesSnapshot,
+			Name:      team.Name,
+			TeamUrl:   team.TeamURL,
+			AvatarUrl: team.AvatarURL,
+		}
+		teamsSnapshot = append(teamsSnapshot, teamSnapshot)
+	}
+
+	slices.SortStableFunc(teamsSnapshot, func(a, b *snapshotpb.Team) bool {
+		return a.Id < b.Id
+	})
+
+	payload := &snapshotpb.Payload{
+		Data: &snapshotpb.Payload_Teams{
+			Teams: &snapshotpb.Teams{
+				Teams: teamsSnapshot,
+			},
+		},
+	}
+
+	serialisedTeams, err := proto.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("could not serialize teams payload: %w", err)
+	}
+
+	e.serialisedTeams = serialisedTeams
+
+	return nil
+}
+
+func (e *SnapshotEngine) serialiseTeamSwitches() error {
+	teamSwitches := e.Engine.teamSwitches
+	teamSwitchesSnapshot := make([]*snapshotpb.TeamSwitch, 0, len(teamSwitches))
+
+	for partyID, teamSwitch := range teamSwitches {
+		teamSwitchSnapshot := &snapshotpb.TeamSwitch{
+			FromTeamId: string(teamSwitch.fromTeam),
+			ToTeamId:   string(teamSwitch.toTeam),
+			PartyId:    string(partyID),
+		}
+		teamSwitchesSnapshot = append(teamSwitchesSnapshot, teamSwitchSnapshot)
+	}
+
+	slices.SortStableFunc(teamSwitchesSnapshot, func(a, b *snapshotpb.TeamSwitch) bool {
+		return a.PartyId < b.PartyId
+	})
+
+	payload := &snapshotpb.Payload{
+		Data: &snapshotpb.Payload_TeamSwitches{
+			TeamSwitches: &snapshotpb.TeamSwitches{
+				TeamSwitches: teamSwitchesSnapshot,
+			},
+		},
+	}
+
+	serialisedTeamSwitches, err := proto.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("could not serialize team switches payload: %w", err)
+	}
+
+	e.serialisedTeamSwitches = serialisedTeamSwitches
+
+	return nil
+}
+
+func (e *SnapshotEngine) loadTeams(p *types.Payload, teams []*snapshotpb.Team) error {
+	e.Engine.loadTeamsFromSnapshot(teams)
+
+	serialisedTeams, err := proto.Marshal(p.IntoProto())
+	if err != nil {
+		return fmt.Errorf("could not serialize teams payload: %w", err)
+	}
+
+	e.serialisedTeams = serialisedTeams
```

Comment:
any reason we keep all the serialised state around? I could not see where we reuse it ever?

Reply:
True, I blindly reproduced what we do everywhere. I think this things should disappear. Some project-wide cleanup should be done regarding this. This is probably a left-over from the initial snapshot implementation

Reply:
Removed.